src/pollen.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Pollen charm business logic."""
+
+import glob
+import subprocess
+from pathlib import Path
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+from charms.operator_libs_linux.v2 import snap
+
+from exceptions import ConfigurationWriteError, InstallError
+
+SNAP_NAME = "gtrkiller-pollen"
+
+
+class PollenService:
+    """Pollen service class."""
+
+    def prepare(self, unit_name, charm_state) -> None:
+        """Install packages and write configuration files.
+
+        Args:
+            unit_name: ¨Pollen charm's unit name.
+            charm_state: Pollen charm's CharmState instance.
+
+        Raises:
+            InstallError: if the packages fail to install
+            ConfigurationWriteError: something went wrong writing the configuration
+        """
+        try:
+            apt.update()
+            snap.add(SNAP_NAME, channel="candidate")
+        except FileNotFoundError as exc:
+            raise InstallError from exc
+        try:
+            subprocess.run(
+                [
+                    "rsync",
+                    f"/var/lib/juju/agents/unit-{unit_name}/charm/files/logrotate.conf",
+                    "/etc/logrotate.d/pollen",
+                ],
+                check=True,
+            )
+            subprocess.run(
+                [
+                    "rsync",
+                    f"/var/lib/juju/agents/unit-{unit_name}/charm/files/rsyslog.conf",
+                    "/etc/rsyslog.d/40-pollen.conf",
+                ],
+                check=True,
+            )
+            systemd.service_restart("rsyslog.service")
+        except (FileNotFoundError, systemd.SystemdError) as exc:
+            raise ConfigurationWriteError from exc
+        if (
+            glob.glob("/dev/tpm*")
+            or Path("/sys/class/misc/hw_random/rng_available").read_text(encoding="utf-8")
+            != "tpm-rng-0"
+        ):
+            try:
+                apt.add_package("rng-tools5")
+                self.check_rng_file(charm_state)
+                systemd.service_restart("rngd.service")
+            except FileNotFoundError as exc:
+                raise ConfigurationWriteError from exc
+
+    def start(self):
+        """Start the pollen service."""
+        cache = snap.SnapCache()
+        pollen = cache[SNAP_NAME]
+        pollen.start()
+
+    def stop(self):
+        """Stop the pollen service."""
+        cache = snap.SnapCache()
+        pollen = cache[SNAP_NAME]
+        pollen.stop()
+
+    def check_rng_file(self, charm_state):
+        """Check if the rng-tools-debian file needs modification.
+
+        Args:
+            charm_state: Pollen charm's CharmState instance.
+        """
+        file = Path("/etc/default/rng-tools-debian")
+        charm_state.rng_tools_file = (
+            None
+            if not file.exists()
```

Comment:
In addition, perhaps worth a comment? What is 2 here?

Reply:
The file already has this line by default commented, so we should check if it appears in the file twice (commented and actually written)

Reply:
Worth a comment? With perhaps a bit more detail - I still don't fully follow why it is commented any why it matters 🙂

Reply:
From my experience, the rngd service will fail if the file does not contain the default commented contents

Reply:
You're checking for two entries, not one commented and one uncommented entry. I agree this seems involved enough to be better handled as suggested in the comment thread below, but I think when you do we should be careful to actually check for one commented entry and one non-commented entry if that's really what's needed (it would be good to see what happens if that commented entry doesn't exist but the uncommented one does, as that sounds very odd that the code would be looking for a comment specifically).